cmd/dagger/functions.go

```diff
@@ -0,0 +1,546 @@
+			q:      querybuilder.Query(),
+			c:      dag.GraphQLClient(),
+			e:      engineClient,
+			mod:    modDef,
+			loader: vtx,
+		}
+
+		help, _ := cmd.Flags().GetBool("help")
+
+		if fc.Execute != nil {
+			if help {
+				return cmd.Help()
+			}
+			return fc.Execute(c, cmd)
+		}
+
+		// Select constructor
+		c.Select(obj.Name)
+
+		// Add main object's functions as subcommands
+		fc.addSubCommands(ctx, dag, obj, c, cmd)
+
+		if help {
+			return cmd.Help()
+		}
+
+		subCmd, subArgs, err := cmd.Find(cmd.Flags().Args())
+		if err != nil {
+			return fmt.Errorf("find command: %w", err)
+		}
+
+		if subCmd.Name() == cmd.Name() {
+			if len(subArgs) > 0 {
+				return fmt.Errorf("unknown command %q for %q", subArgs[0], cmd.CommandPath())
+			}
+			return cmd.Help()
+		}
+
+		return fc.run(subCmd, subArgs)
+	})
+}
+
+func (*FuncCommand) run(cmd *cobra.Command, args []string) error {
+	err := cmd.RunE(cmd, args)
+	if err != nil {
+		return fmt.Errorf("execute command %q: %w", cmd.Name(), err)
+	}
+	return nil
+}
+
+func (fc *FuncCommand) addSubCommands(ctx context.Context, dag *dagger.Client, obj *modObject, c *callContext, cmd *cobra.Command) {
+	if obj != nil {
+		for _, fn := range obj.Functions {
+			subCmd := fc.makeSubCmd(ctx, dag, fn, c)
+			cmd.AddCommand(subCmd)
+		}
+		for _, field := range obj.Fields {
+			subCmd := fc.makeSubCmd(ctx, dag, &modFunction{Name: field.Name, Description: field.Description, ReturnType: field.TypeDef}, c)
+			cmd.AddCommand(subCmd)
+		}
+	}
+}
+
+func (fc *FuncCommand) makeSubCmd(ctx context.Context, dag *dagger.Client, fn *modFunction, c *callContext) *cobra.Command {
+	newCmd := &cobra.Command{
+		Use:   cliName(fn.Name),
+		Short: fn.Description,
+		RunE: func(cmd *cobra.Command, args []string) (err error) {
+			cmd.SetContext(ctx)
+			rec := progrock.FromContext(ctx)
+
+			c.mod.LoadObject(fn.ReturnType)
+
+			for _, arg := range fn.Args {
+				c.mod.LoadObject(arg.TypeDef)
+			}
+
+			argValues := make(map[string]any)
+
+			for _, arg := range fn.Args {
+				p, err := arg.AddFlag(cmd.Flags(), dag)
+				if err != nil {
+					return fmt.Errorf("set flag %q: %w", arg.FlagName(), err)
+				}
+				argValues[arg.Name] = p
+
+				// TODO: This won't work on its own because cobra checks required
+				// flags before the RunE function is called. We need to manually
+				// check the flags after parsing them.
+				if !arg.TypeDef.Optional {
+					cmd.MarkFlagRequired(arg.FlagName())
+				}
+			}
+
+			if fc.OnSetFlags != nil {
+				return fc.OnSetFlags(cmd, fn.ReturnType)
+			}
+
+			if err := cmd.ParseFlags(args); err != nil {
+				return fmt.Errorf("parse sub-command flags: %w", err)
+			}
+
+			fc.addSubCommands(ctx, dag, fn.ReturnType.AsObject, c, cmd)
+
+			// Need to make the query selection before chaining off.
+			returnType, err := fc.selectFunc(fn, c, argValues, dag)
+			if err != nil {
+				return fmt.Errorf("query selection: %w", err)
+			}
+
+			help, _ := cmd.Flags().GetBool("help")
```

Comment:
I agree that it should be handled in the CLI. For sub-commands we can just limit to `-h`. In fact, the shorthand notation is a simple way to avoid conflict with function arguments for other cases as well, for example when returning a `Directory` and we want the `--export-path`. So we need a consistent way to add these flags for extra arguments without conflicting with the user's arguments.